libexec/rtld-elf/rtld_c18n.h

```diff
@@ -0,0 +1,225 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifndef RTLD_C18N_H
+#define RTLD_C18N_H
+
+#include <stdint.h>
+
+/*
+ * Global symbols
+ */
+#define	C18N_FUNC_SIG_COUNT	72
+
+extern uintptr_t sealer_pltgot, sealer_tramp;
+extern const char *ld_compartment_utrace;
+extern const char *ld_compartment_enable;
+extern const char *ld_compartment_overhead;
+extern const char *ld_compartment_sig;
+
+void tramp_init(void);
+
+/*
+ * Policies
+ */
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+#define	C18N_RTLD_COMPARTMENT_ID	1
+#else
+#define	C18N_RTLD_COMPARTMENT_ID	0
+#endif
+#define	C18N_COMPARTMENT_ID_MAX	(UINT16_MAX >> 1)
+
+typedef uint16_t compart_id_t;
+
+struct compart {
+	/*
+	 * Name of the compartment
+	 */
+	const char *name;
+	/*
+	 * NULL-terminated array of libraries that belong to the compartment
+	 */
+	const char **libraries;
+};
+
+struct policy {
+	struct compart *coms;
+	size_t count;
+};
+
+void tramp_add_comparts(struct policy *);
+compart_id_t compart_id_allocate(const char *);
+
+/*
+ * Stack switching
+ */
+struct stk_table {
+	union {
+		void *(*resolver)(unsigned);
+		SLIST_ENTRY(stk_table) next;
+	};
+	size_t capacity;
+	struct stk_table_stack {
+		void *bottom;
+		size_t size;
+	} stacks[];
+};
+
+struct Struct_Stack_Entry {
+    SLIST_ENTRY(Struct_Stack_Entry) link;
+    void *stack;
+};
+
+void allocate_stk_table(void);
+void *_rtld_get_rstk(unsigned);
+
+static inline unsigned
+compart_id_to_index(compart_id_t cid)
+{
+	struct stk_table dummy;
+
+	return (sizeof(*dummy.stacks) * cid / sizeof(dummy.stacks->bottom));
+}
+
+static inline struct stk_table *
+stk_table_get(void)
+{
+	struct stk_table *table;
+
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	asm ("mrs	%0, rctpidr_el0" : "=C" (table));
+#else
+	asm ("mrs	%0, ctpidr_el0" : "=C" (table));
+#endif
+	return (table);
+}
+
+static inline void
+stk_table_set(struct stk_table *table)
+{
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	asm ("msr	rctpidr_el0, %0" :: "C" (table));
+#else
+	asm ("msr	ctpidr_el0, %0" :: "C" (table));
+#endif
+}
+
+static inline void
+trusted_stk_set(void *sp)
+{
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	asm ("msr	rcsp_el0, %0" :: "C" (sp));
+#else
+	asm ("mov	csp, %0" :: "C" (sp));
```

Comment:
This is not a thing you can easily just do like that in C as it'll totally screw up the compiler. I suggest you just don't define (un)trusted_stk_set and only define the one that uses RCSP_EL0, since that's the only function you actually need of the two. Whichever one sets CSP is unused and it needs to stay that way unless you want explosions.